test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -132,3 +142,140 @@
                        :breakout    [[:expression "expr"]]}]]]
               (testing (format "%s function works as expected on %s column for driver %s" op col-type driver/*driver*)
                 (is (= (set expected) (set (test-date-extract query))))))))))))
+
+(defn- date-math
+  [op x amount unit col-type]
+  (let [amount (if (= op :date-add)
+                 amount
+                 (- amount))
+        result (cond
+                 ;; unlike everybody else, vertica decided when adding 1 year it will do that by adding 365 days,
+                 ;; or 90 days if you want to add a quarter.
+                 ;; That means if you do `select date '2004-02-02' + interval '2 year';` in vertica, it will returns
+                 ;; `2006-02-01` because 2004 is a leap year. All others DBs will return `2006-02-02` instead.
+                 (and (= driver/*driver* :vertica) (#{:year :quarter :month} unit))
+                 (u.date/add x :day (* amount (case unit
+                                                :year    365
+                                                :quarter (* 3 30)
+                                                :month   30)))
+                 :else
+                 (u.date/add x unit amount))
+        fmt    (cond
+                 ;; the :date column of :presto should have this format too,
+                 ;; but the test data we created for presto is datetime even if we define it as date
+                 (and (= driver/*driver* :presto) (#{:text-as-date} col-type))
+                 "yyyy-MM-dd"
+
+                 :else
+                 "yyyy-MM-dd HH:mm:ss")]
+    (t/format fmt result)))
+
+(defn- mongo-major-version [db]
+  (-> (get-in db [:details :version])
```

Comment:
I think this should be made resilient to the version string not being present. The `str/split` and `parse-long` will both error out if that version string is missing. If we do want to error we should have a better error message than those will provide.

Reply:
I don't really love the pattern of checking the MongoDB version here. If supporting this stuff is dependent on the Mongo version, we should be using `driver/database-supports?` to check support for the feature rather than `driver/supports?`. That being said we don't have test tooling around that yet -- `mt/normal-drivers-with-feature` checks `driver/supports?` rather than `driver/database-supports?`. So I won't put my foot down on this.

That being said if you want to do this the right way you can always do something like

```clj
(mt/test-drivers (mt/normal-drivers-with-feature :date-arithmetics)
  (when (driver/database-supports? driver/*driver* (mt/db) :date-arithmetics)
    ...))
```

Better yet, you can turn that into a test util macro so we can start using that everywhere.